SYCL/Matrix/joint_matrix_query_general.cpp

```diff
@@ -0,0 +1,254 @@
+         }); // parallel for
+   }).wait();
+}
+
+bool is_this_amx_device(queue q) {
+  auto dev_type =
+      q.get_device().template get_info<sycl::info::device::device_type>();
+  if (dev_type == info::device_type::cpu) {
+    std::cout << "This is a CPU Device" << std::endl;
+    unsigned eax, ebx, ecx, edx;
+
+    eax = 1; /* processor info and feature bits */
+    native_cpuid(&eax, &ebx, &ecx, &edx);
+    int model = (eax >> 4) & 0xF;
+    int family = (eax >> 8) & 0xF;
+    int extended_model = (eax >> 16) & 0xF;
+    int extended_family = (eax >> 20) & 0xFF;
+
+    // SPR processsor means AMX is present
+    if (model == 15 && family == 6 && extended_model == 8 &&
+        extended_family == 0) {
+      std::cout << "This is a SPR CPU Device" << std::endl;
+      return true;
+    }
+  }
+  return false;
+}
+
+bool is_this_xmx16_device(queue q) {
+  auto dev_type =
+      q.get_device().template get_info<sycl::info::device::device_type>();
+  if (dev_type == info::device_type::gpu) {
+    std::cout << "This is a GPU Device" << std::endl;
+
+    auto dev_name =
+        q.get_device().template get_info<sycl::info::device::name>();
+    size_t found = dev_name.find("Intel(R) Graphics [0x0bd5]"); // PVC
+    if (found != std::string::npos) {
+      std::cout << "This is a PVC Device which contains XMX-16 TPU"
+                << std::endl;
+      return true;
+    }
+  }
+  return false;
+}
+
+bool is_this_xmx8_device(queue q) {
+  auto dev_name = q.get_device().template get_info<sycl::info::device::name>();
+  size_t found = dev_name.find("Intel(R) Graphics [0x5610]"); // ATS-P
+  if (found != std::string::npos) {
+    std::cout << "This is a ATS-M Device which contains XMX-8 TPU" << std::endl;
+    return true;
+  }
+  return false;
+}
+
+template <typename T1, typename T2, size_t M, size_t N, size_t K>
+bool query_and_matrix_multiply(big_matrix<T1, M, N> &C, big_matrix<T2, M, K> &A,
+                               big_matrix<T2, K / 4, N * 4> &B) {
+  queue q;
+
+  if (is_this_amx_device(q)) {
+    using myparams2 = tpu_params<tpu::amx, int8_t, int8_t, int>;
+    constexpr int TM = myparams2::M;
+    constexpr int TN = myparams2::N;
+    constexpr int TK = myparams2::K;
+    std::cout << "AMX query default sizes are: M " << TM << " N " << TN << " K "
+              << TK << std::endl;
+    matrix_multiply<T1, T2, M, N, K, TM, TN, TK>(q, C, A, B);
+    return true;
+  }
+
+  if (is_this_xmx16_device(q)) {
+    using myparams2 = tpu_params<tpu::xmx16, int8_t, int8_t, int>;
+    constexpr int TM = myparams2::M;
+    constexpr int TN = myparams2::N;
+    constexpr int TK = myparams2::K;
+    std::cout << "XMX of PVC query default sizes are: M " << TM << " N " << TN
+              << " K " << TK << std::endl;
+    matrix_multiply<T1, T2, M, N, K, TM, TN, TK>(q, C, A, B);
+    return true;
+  }
+
+  if (is_this_xmx8_device(q)) {
+    using myparams2 = tpu_params<tpu::xmx8, int8_t, int8_t, int>;
```

Comment:
you are right this is a typo. I only run the code on SPR (AMX) and PVC (XMX16).
Since I did not run it on ATS-M (that has XMX8), i did not catch the typo. I will run on ATS-M and fix it.